### ISSUE TITLE
[GlobalStep] "400 Bad Request" error triggers at Query Monitor at "Order Recevied" page after clicking on "Resubscribe" button on "Checkout" page.

### ISSUE THREAD

**Issue by gglobalstep** [Automattic/woocommerce-payments#4430]
### Bug Description:
 "400 Bad Request" error triggers at Query Monitor at "Order Recevied" page after clicking on "Resubscribe" button on "Checkout" page.

### Environment:
Woocommerce Version: WooCommerce 6.6.1
WooCommerce  Payments 4.5.0-test-1
WordPress version: v6.0

PC:
Windows 10
Chrome(Version 103.0.5060.114) 
Firefox (Version 102.0.1)
Mac OS:
Safari(Version 15.5)

### Steps To Reproduce:
1. Create any test site using JN site.
1. Install and activate all the required plugins.
1. Upload WooCommerce Payments - 4.5.0-test-1 plugin.
1. Purchase an "Subscription product" on frontend.
1. Navigate to My account → Subscriptions.
1. Click on the "subscription ID" of the latest purchased "Subscription product".
1. Click on the "Cancel" button.
1. Confirm the status is "Pending Cancellation".
1. Again go to the My account → Subscriptions.
1. Click on the same "Subscription product".
1. Click on the "Resubscribe" button.
1. Complete checkout with any testing card (e.g. 4242424242424242).
1. Observe that "400 Bad Request" error triggers at Query Monitor at "Order Recevied" page after clicking on "Resubscribe" button on "Checkout" page.

### Actual Result:
 "400 Bad Request" error triggers at Query Monitor at "Order Recevied" page after clicking on "Resubscribe" button on "Checkout" page.

### Expected Result:
Order should be recevied without any error after clicking on "Resubscribe" button on "Checkout" page.

### Screenshot:

![#4430](https://user-images.githubusercontent.com/41110392/178522296-4d3faf5c-ed6d-436d-8af1-bae1297c3c55.jpg)

*Isolating the problem (mark completed items with an [x]):*
- [ ] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [ ] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [x] I can reproduce this bug consistently using the steps above.

<details>

`
### WordPress Environment ###

WC Version: 6.6.1
REST API Version: ✔ 6.6.1
WC Blocks Version: ✔ 7.6.2
Action Scheduler Version: ✔ 3.4.0
Log Directory Writable: ✔
WP Version: 6.0
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: ✔
WP Cron: ✔
Language: en_US
External object cache: –

### Server Environment ###

Server Info: Apache/2.4.54 (Unix) OpenSSL/1.0.2g
PHP Version: 7.4.30
PHP Post Max Size: 1 GB
PHP Time Limit: 30
PHP Max Input Vars: 5000
cURL Version: 7.47.0
OpenSSL/1.0.2g

SUHOSIN Installed: –
MySQL Version: 5.7.33-0ubuntu0.16.04.1-log
Max Upload Size: 512 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 6.6.1
WC Database Prefix: wp_
Total Database Size: 7.57MB
Database Data Size: 5.44MB
Database Index Size: 2.13MB
wp_woocommerce_sessions: Data: 0.05MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_order_items: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_order_itemmeta: Data: 0.09MB + Index: 0.09MB + Engine InnoDB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_actions: Data: 0.16MB + Index: 0.14MB + Engine InnoDB
wp_actionscheduler_claims: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_groups: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_logs: Data: 0.13MB + Index: 0.11MB + Engine InnoDB
wp_commentmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_comments: Data: 0.14MB + Index: 0.09MB + Engine InnoDB
wp_links: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_options: Data: 3.48MB + Index: 0.08MB + Engine InnoDB
wp_postmeta: Data: 0.45MB + Index: 0.47MB + Engine InnoDB
wp_posts: Data: 0.13MB + Index: 0.06MB + Engine InnoDB
wp_snippets: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_termmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_terms: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_term_relationships: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_usermeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_users: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_admin_notes: Data: 0.05MB + Index: 0.00MB + Engine InnoDB
wp_wc_admin_note_actions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_category_lookup: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_customer_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_coupon_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_product_lookup: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_stats: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_order_tax_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_product_attributes_lookup: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_download_directories: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_meta_lookup: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_wc_rate_limits: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_reserved_stock: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_tax_rate_classes: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB + Engine InnoDB

### Post Type Counts ###

attachment: 25
page: 10
post: 2
product: 22
product_variation: 7
revision: 3
shop_coupon: 1
shop_order: 68
shop_order_refund: 8
shop_subscription: 25
wp_global_styles: 1

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ✔

### Active Plugins (6) ###

Query Monitor: by John Blackbourn – 3.9.0
Code Snippets: by Code Snippets Pro – 3.1.1
Companion Plugin: by Osk – 1.27
WooCommerce Payments Dev Tools: by Automattic –
WooCommerce Payments: by Automattic – 4.5.0-test-1
WooCommerce: by Automattic – 6.6.1

### Inactive Plugins (5) ###

Akismet Anti-Spam: by Automattic – 4.2.5
Hello Dolly: by Matt Mullenweg – 1.7.2
Jetpack: by Automattic – 11.1.1
WooCommerce Blocks: by Automattic – 8.0.0
WooCommerce Subscriptions: by WooCommerce – 4.4.0

### Dropin Plugins (1) ###

db.php: Query Monitor Database Class (Drop-in)

### Settings ###

API Enabled: –
Force SSL: –
Currency: USD ($)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
subscription (subscription)
variable (variable)
variable subscription (variable-subscription)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: –
Enforce Approved Product Download Directories: ✔

### WC Pages ###

Shop base: #6 - /shop/
Cart: #7 - /cart/
Checkout: #8 - /checkout/
My account: #9 - /my-account/
Terms and conditions: ❌ Page not set

### Theme ###

Name: Storefront
Version: 4.1.1
Author URL: https://woocommerce.com/
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ✔

### Templates ###

Overrides: –

### Subscriptions ###

WCS_DEBUG: ✔ No
Subscriptions Mode: ✔ Live
Subscriptions Live URL: https://old-gentoo.jurassic.ninja
Subscription Statuses: wc-pending: 1
wc-cancelled: 4
wc-active: 20

WooCommerce Account Connected: ❌ No

### Store Setup ###

Country / State: United States (US) — California

### Subscriptions by Payment Gateway ###

WooCommerce Payments: wc-active: 20
wc-cancelled: 4
wc-pending: 1

### Payment Gateway Support ###

WooCommerce Payments: products
refunds
multiple_subscriptions
subscription_cancellation
subscription_payment_method_change_admin
subscription_payment_method_change_customer
subscription_payment_method_change
subscription_reactivation
subscription_suspension
subscriptions
gateway_scheduled_payments
tokenization
add_payment_method

### Admin ###

Enabled Features: activity-panels
analytics
coupons
customer-effort-score-tracks
experimental-products-task
experimental-import-products-task
experimental-fashion-sample-products
homescreen
marketing
mobile-app-banner
navigation
onboarding
onboarding-tasks
remote-inbox-notifications
remote-free-extensions
payment-gateway-suggestions
shipping-label-banner
subscriptions
store-alerts
transient-notices
wc-pay-promotion
wc-pay-welcome-page
wc-pay-subscriptions-page

Disabled Features: minified-js
settings

Daily Cron: ✔ Next scheduled: 2022-07-13 05:55:27 +00:00
Options: ✔
Notes: 42
Onboarding: completed

### WooCommerce Payments ###

Version: 4.5.0-test-1
Connected to WPCOM: Yes
Blog ID: 208780402
Account ID: acct_1LKcCb2HOltEgIvg

### Action Scheduler ###

Complete: 385
Oldest: 2022-07-12 06:10:26 +0000
Newest: 2022-07-12 11:18:56 +0000

Pending: 26
Oldest: 2022-07-12 15:04:30 +0000
Newest: 2022-09-12 10:36:35 +0000

### Status report information ###

Generated at: 2022-07-12 15:02:24 +00:00
`

</details>

**Comment by htdat** [comment]
### Some relevant links: 

- eported by GlobalStep p1657627660886829-slack-C011ENB20Q1 (this includes creds for the test site)
- Gamma discussed it here too p1657639013495069/1657609429.947819-slack-C01BPL3ALGP 

### Here is what I found out 

#### tl;dr: 

It seems something wrong specifically with subscriptions when setup intents (`seti_`) were provided for `intented_id` paramer in `generate_payment_success_note`. It's expected that this is a payment intent (`pi_`).

https://github.com/Automattic/woocommerce-payments/blob/f4ef1de411128245f8cd617e088f4803fca8b297/includes/class-wc-payments-order-service.php#L335


#### Logs 

- WCPay log under wp-content/uploads/wc-logs and wp-content/debug.log for this site: 2ce1d-pb (I am saving it in a private paste bin as the test site will be deleted in 7 days and it includes some secrete info)

#### Details

With the provided error above, I could find out the relevant part for this setup intent ID `seti_1LKf8P2HOltEgIvg81ZaHF3H` in the wc-logs file: 

```
2022-07-12T09:12:18+00:00 INFO ENVIRONMENT: array (
  'WP_User' => 'demo',
  'HTTP_REFERER' => 'https://old-gentoo.jurassic.ninja/checkout/',
  'HTTP_USER_AGENT' => 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/103.0.0.0 Safari/537.36',
  'REQUEST_URI' => '/checkout/order-received/101/?key=wc_order_xxxxxxxx&order_id=101&wc_payment_method=woocommerce_payments&_wpnonce=0892be7b46&save_payment_method=yes&setup_intent=seti_1LKf8P2HOltEgIvg81ZaHF3H&setup_intent_client_secret=seti_xxxxxxxxxx&redirect_status=succeeded',
  'DOING_AJAX' => false,
  'DOING_CRON' => false,
  'WP_CLI' => false,
)
2022-07-12T09:12:18+00:00 INFO REQUEST GET https://public-api.wordpress.com/wpcom/v2/sites/%s/wcpay/timeline/seti_1LKf8P2HOltEgIvg81ZaHF3H?test_mode=1
2022-07-12T09:12:18+00:00 INFO HEADERS: array (
  'Content-Type' => 'application/json; charset=utf-8',
  'User-Agent' => 'WooCommerce Payments/4.5.0-test-1',
)
2022-07-12T09:12:18+00:00 ERROR Invalid intent ID (wcpay_bad_request)
2022-07-12T09:12:18+00:00 INFO Can not generate the detailed note for intent_id seti_1LKf8P2HOltEgIvg81ZaHF3H. Reason: Error: Invalid intent ID
```

There is no fatal error in debug.log file, and with this log entry `Can not generate the detailed note for intent_id seti_1LKf8P2HOltEgIvg81ZaHF3H`, it indicates that the `try-catch` part in `generate_payment_success_note` is handled properly. 

**But the issue is why the setup intent ID is provided here.** 

In this order https://old-gentoo.jurassic.ninja/wp-admin/post.php?post=101&action=edit, the note looks like: 

![Markup on 2022-07-13 at 13:05:21](https://user-images.githubusercontent.com/10045087/178662200-31e1ef48-9683-4bb0-a81b-480886163dba.png)

HTML for this note:

```
A payment of <span class="woocommerce-Price-amount amount"><span class="woocommerce-Price-currencySymbol">$</span>0.00</span> USD was <strong>successfully charged</strong> using WooCommerce Payments (<code></code>).
``` 

The `<code></code>` here indicate that `$transaction_url` is empty, and then trace it back, it means `$charge_id` is empty too.

IMHO, the cause for this is similar to the issue with the setup intent ID above.

**Comment by htdat** [comment]
👋 Hi @Automattic/helix, can you look at this issue? It might be some changes on WC Subscription might create this (regression?) issue?

**Comment by haszari** [comment]
@htdat  I couldn't reproduce this. I was testing on my local dev (docker) with latest WooCommerce and current `develop` branch (22966094e33997eaca02be0bc4a542bdb2c28a31). Did you reproduce this on a clean site, or is it possibly related to some specific state or config on the Jurassic Ninja site?

**Comment by htdat** [comment]
@haszari  I did not try to replicate the error when I wrote my comment above. But @jbordonado (current release lead) can reproduce the error too. 

I've tried to reproduce this error, and here is what I found out: 

- In step 12, if I use a saved card, there is no error.
- However, if I use a totally new card (this is also what it means for step 12), the error happens. 

I think that makes sense with a setup intent here. Can you try again?

**Comment by haszari** [comment]
> In step 12, if I use a saved card, there is no error.

Thanks @htdat. that was probably my issue – my site has a bunch of saved cards from previous testing, I'll clear them all out :)

**Comment by haszari** [comment]
I haven't had a chance to follow up on this – @mattallan (current reviewer) could you try reproducing the 400 error? Keen to confirm if there's something incorrect happening with setup intent as per @htdat 's diagnosis above.

**Comment by mattallan** [comment]
Hi @haszari and @htdat, I've been able to reproduce this in 4.5.0-test-1 but this is not a new issue. This issue can also be reproduced on `4.2.0-test-2` which is when this error started appearing.

Here's the comment that introduced the issue: https://github.com/Automattic/woocommerce-payments/commit/a10a619f263cc86771e48326a6b773245822aa9e

I tested the latest `4.5.0-test-2` zip and can confirm the error is no longer present.
Out of interest, I also tested regular tokenized subscriptions (with WC Subscriptions active) using similar testing instructions above and it seems this issue only affects WCPay Subscriptions.

Not sure if this helps isolate the problem though 😅

**Comment by htdat** [comment]
> I tested the latest `4.5.0-test-2` zip and can confirm the error is no longer present.

It's because of this PR https://github.com/Automattic/woocommerce-payments/pull/4429, which dispatched the process of creating fee detail notes into an async job. But the underlying issue is that a setup intent should not be provided for `$intent_id` param of the following function. It needs to be a payment intent: 

https://github.com/Automattic/woocommerce-payments/blob/f4ef1de411128245f8cd617e088f4803fca8b297/includes/class-wc-payments-order-service.php#L335

**Comment by mattallan** [comment]
I've spent a bit more time looking into this issue and trying to answer your question which is: **Why is a `setup_intent` being passed to `generate_payment_success_note()` when it's expected to be a `payment_intent`?**

The answer to this question is because when you hit "Resubscribe" on a subscription that has pending cancellation status, the checkout is $0. In the `process_redirect_payment()` method we check if the order is $0 and if it is, we get the `setup_intent`. See relevant code [here](https://github.com/Automattic/woocommerce-payments/blob/09ec73bc5fd0a0b0fd6c264b220abd337301e51d/includes/payment-methods/class-upe-payment-gateway.php#L611-L621).

Why is the checkout $0? This is because a pending cancellation subscription is still considered "active" until the pre-paid term (or end date) has been met. So the customer shouldn't need to pay anything upfront and the first payment of the new subscription will be set to the end date of their existing subscription that is set to pending cancellation.

Screenshot of the $0 checkout where the API errors can be seen:
![image](https://user-images.githubusercontent.com/2275145/180361051-e0f41cab-c51f-45ad-b862-be6fd6e4bc1d.png)

Hope this makes sense!

In my opinion there's no bug/issue with WCPay Subscriptions and the resubscribe flow. Having a `setup_intent` is correct isn't it? Considering the checkout is $0?

**Comment by haszari** [comment]
Closing based on @mattallan analysis above. If this is causing problems for merchants, please reopen! FYI @htdat

**Comment by htdat** [comment]
Reopening this issue as this error is mentioned again in p1677659325507429/1677646041.656859-slack-C01BPL3ALGP, and moving it to Gamma backlog queue. 

Idea to fix the issue or avoid logging an error. 

- Create is_valid_payment_intent_id as a util function like this: 

```php
   private function is_valid_payment_intent_id( $payment_intent_id ) {
        return 1 === preg_match( '/^pi_\w+$/', $payment_intent_id );
    }
```

- Add this check and return early for add_fee_breakdown_to_order_notes
https://github.com/Automattic/woocommerce-payments/blob/23141d7f5d4a513d3e815366afe8d5542ca57e2e/includes/class-wc-payments-order-service.php#L394-L394

- Or better to return even earlier and avoid queuing this action
https://github.com/Automattic/woocommerce-payments/blob/0d19887326fab36e5e550d53e41e63d60c22e4cf/includes/class-wc-payments-order-service.php#L1473

**Comment by kalessil** [comment]
Seems to be subscription related, @diegocurbelo can you please look into the issue? We ar not sure that's something related to fees.

**Comment by vbelolapotkov** [comment]
Routing it to subs focus based on @mattallan's comment here: https://github.com/Automattic/woocommerce-payments/issues/4430#issuecomment-1192171929

I don't think resubscribe button leading to zero amount checkout is correct behavior.

**Comment by flakas** [comment]
Had another look at this as part of a backlog review — it still happens on 11.0.0. I can see the error in the logs:

```
GET .../wcpay/timeline/seti_1U6yXpCIyYVln6qcEQHoKByy?test_mode=1
Error  Invalid intent ID (wcpay_bad_request)
Info   Can not generate the detailed note for intent_id seti_…  Reason: Invalid intent ID
```

In my tests it only fires when the $0 resubscribe is paid with a *new* card, because that's the only case where we create a setup intent. Pay with a saved card, or have "$0 Initial Checkout" enabled, and nothing happens at all — no intent, no error. The Resubscribe button also only appears on Stripe Billing stores, so the path isn't reachable otherwise.

When it does fire, we hand the setup intent ID to code expecting a payment intent and get back `Invalid intent ID`. Same ID also goes to the intentions endpoint, and ends up stored as the order's transaction ID — which is what's behind #6577's broken transaction link on $0 orders. Worth fixing all three together rather than one at a time.

Impact appears to be low: the error is caught and logged, checkout completes, the customer is fine. What's lost is the fee breakdown note on the order.

Leaving this as ready for engineering.
